daemons/attrd/attrd_cib.c

```diff
@@ -45,6 +45,87 @@ attrd_cib_destroy_cb(gpointer user_data)
     }
 }
 
+/*!
+ * \internal
+ * \brief Check a patchset change for deletion of node attribute values
+ *
+ * \param[in] xml   Patchset change element
+ * \param[in] data  Ignored
+ *
+ * \return pcmk_rc_ok (to always continue to next patchset change)
+ */
+static int
+drop_values_in_deletion(xmlNode *xml, void *data)
+{
+    const char *value = NULL;
+    GRegex *regex = NULL;
+    GMatchInfo *match_info = NULL;
+    gchar *node_id = NULL;
+    gchar *set_type = NULL;
+    gchar *set_id = NULL;
+    gchar *attr_id = NULL;
+
+    // Skip this change if it does not look like a deletion
+    value = pcmk__xe_get(xml, PCMK_XA_OPERATION);
+    if (!pcmk__str_eq(value, "delete", pcmk__str_none)) {
+        return pcmk_rc_ok;
+    }
+
+    value = pcmk__xe_get(xml, PCMK_XA_PATH);
+    if (value == NULL) {
+        crm_warn("Ignoring malformed deletion in "
+                 "CIB change notification: No " PCMK_XA_PATH);
+        return pcmk_rc_ok;
+    }
+
+    regex = g_regex_new("^/" PCMK_XE_CIB
+                        "/" PCMK_XE_STATUS
+                        "/" PCMK__XE_NODE_STATE
+                        "\\[@" PCMK_XA_ID "='(?<NODE_ID>[^']+)'\\]"
+                        "(?:"
+                            "/" PCMK__XE_TRANSIENT_ATTRIBUTES
+                            "\\[@" PCMK_XA_ID "='[^']+'\\]"
+                            "(?:"
+                                "/(?<SET_TYPE>[^[/]+)"
+                                "\\[@" PCMK_XA_ID "='(?<SET_ID>[^']+)'\\]"
+                                "(?:"
+                                    "/" PCMK_XE_NVPAIR
+                                    "\\[@" PCMK_XA_ID "='(?<ATTR_ID>[^']+)'\\]"
+                                ")?"
+                            ")?"
+                        ")?$",
+                        0, 0, NULL);
+
+    if (!g_regex_match(regex, value, 0, &match_info)) {
+        goto done;
+    }
+
+    node_id = g_match_info_fetch_named(match_info, "NODE_ID");
+    set_type = g_match_info_fetch_named(match_info, "SET_TYPE");
+    set_id = g_match_info_fetch_named(match_info, "SET_ID");
+    attr_id = g_match_info_fetch_named(match_info, "ATTR_ID");
+
+    if (!pcmk__str_empty(attr_id)) {
+        attrd_drop_removed_value(set_type, attr_id);
```

Comment:
It doesn't hurt to have `attrd_drop_removed_value()` and `attrd_drop_removed_set()`. But upon re-reading the commit message, I am starting to wonder whether they're necessary. It seems like the only situations that we needed to deal with in this PR are those where a node is purged. In those cases, the node's whole `<transient_attributes>` element is deleted.

I don't think there should be a situation where we have to use this "set to `NULL` and delete from CIB later" trick for deleting single attributes or attribute sets -- at least based on the scenario described in the commit message.

I'm on the fence about whether we should keep these due to paranoia, or drop them in order to significantly simplify the regex and the support code. Definitely open to discussion.